README.md
# screeps-modules

[![Build status][travis-ci-badge]][travis-ci] [![NPM][npm-badge]][npm] ![Node version][node-badge] [![bitHound Dependencies][bithound-dependencies-badge]][bithound-dependencies] [![bitHound Dev Dependencies][bithound-dev-dependencies-badge]][bithound-dev-dependencies]  [![Coverage][codecov-badge]][codecov]

A thin client for committing/retrieving modules from a Screeps server, including [private servers](http://support.screeps.com/hc/en-us/articles/213625765-Screeps-private-server-released-). If you are using Webpack, see [langri-sha/screeps-webpack-plugin](https://github.com/langri-sha/screeps-webpack-plugin).

The client uses the [Screeps Web API](http://support.screeps.com/hc/en-us/articles/203022612-Committing-scripts-using-direct-API-access). To authenticate, make sure you have an [account password configured](https://screeps.com/a/#!/account), if you've registered via Steam or GitHub. If you provide a token, it will be used instead to authorize requests.

On private servers, make sure that you have the  [authorization mod](https://github.com/screepsmods/screepsmod-auth) installed and an account created.

For something more comprehensive, check out [screepers/node-sreeps-api](https://github.com/screepers/node-screeps-api).

# Install

```
npm install screeps-modules
```

## Usage

```
import ScreepsModules from 'screeps-modules'

const client = new ScreepsModules(options)
```

### Options

```
const client = new ScreepsModules({
  email: 'EMAIL',
  password: 'PASSWORD',
  token: 'TOKEN'
  serverUrl: 'https://screeps.com',
  serverPassword: 'SERVER_PASS',
  gzip: false
}
```

### `ScreepsModules#commit([branch,] modules): Object`

Commit modules to the provided branch, or to the one that's active in the world.

```
// Update branch 'sim'
await client.commit('sim', {
  main: 'module.exports = () => {console.log(Game.time)}'
})
// => {ok: 1}
// Update branch 'sim'

// Update the active world branch
await client.commit({
  main: 'module.exports = () => {console.log(Game.time)}'
})
// => {ok: 1}
```

### `ScreepsModules#fetch([branch])`

Retrieve scripts from the provided branch, or from the one that's active in the world.

```
// Retrieve branch 'sim'
await client.retrieve('sim')
// => {main: 'module.exports = () => {console.log(Game.time)}'}

// Fetch active world branch
await client.retrieve()
// => Object
```

### `ScreepsModules#refreshToken()`

Updates the client token. Returns the token.

```
await client.refreshToken()
// => 'TOKEN'
```

## Troubleshooting

To avoid leaking sensitive information, the client only returns opaque responses and is safe to use in CI. If you need to inspect the responses, start your process with `DEBUG=screeps-modules node $ARGS`.

[travis-ci]: https://travis-ci.org/langri-sha/screeps-modules
[travis-ci-badge]: https://travis-ci.org/langri-sha/screeps-modules.svg?branch=master
[npm]: https://www.npmjs.com/package/screeps-modules
[npm-badge]: https://img.shields.io/npm/v/screeps-modules.svg
[node-badge]: https://img.shields.io/node/v/screeps-modules.svg
[codecov]: https://codecov.io/gh/langri-sha/screeps-modules
[bithound-dependencies]: https://www.bithound.io/github/langri-sha/screeps-modules/master/dependencies/npm
[bithound-dependencies-badge]: https://www.bithound.io/github/langri-sha/screeps-modules/badges/dependencies.svg
[bithound-dev-dependencies]: https://www.bithound.io/github/langri-sha/screeps-modules/master/dependencies/npm
[bithound-dev-dependencies-badge]: https://www.bithound.io/github/langri-sha/screeps-modules/badges/devDependencies.svg
[codecov-badge]: https://codecov.io/gh/langri-sha/screeps-modules/branch/master/graph/badge.svg


CHANGELOG.md
# Change Log

## [1.2.0]
### Added
- `X-Server-Password` header with corresponding `serverPassword` field for protected private servers

### Changed
- Upgraded dependencies

## [1.1.2] - 2016-11-18
### Added
- Invalidate token on 401 responses
- Always set token found response headers
- Improved error handling

## [1.0.2] - 2016-11-13
### Added
- Coverage
- Token support
- Optional support for compressing commit sources
- Support for fetching modules
- Support for committing modules

[Unreleased]: https://github.com/langri-sha/screeps-modules/commits/master
[1.2.0]: https://github.com/langri-sha/screeps-scripts/compare/v1.1.2...v1.2.0
[1.1.2]: https://github.com/langri-sha/screeps-scripts/compare/v1.0.2...v1.1.2
[1.0.2]: https://github.com/langri-sha/screeps-scripts/compare/283a050...v1.0.2


test.js
import test from 'ava'
import nock from 'nock'

import ScreepsModules from './index'

const ok = [200, {ok: 1}]

test.before(() => {
  nock.disableNetConnect()
})

test('Test commit with defaults', async t => {
  t.plan(2)

  nock('https://screeps.com')
    .matchHeader('content-type', 'application/json')
    .matchHeader('accept', 'application/json')
    .post('/api/user/code', {
      branch: 'default',
      modules: {
        main: 'module.exports = () => {}'
      }
    })
    .basicAuth({
      user: 'foobar',
      pass: 'barbaz'
    })
    .reply(() => {
      t.pass()

      return ok
    })

  const client = await new ScreepsModules({
    email: 'foobar',
    password: 'barbaz'
  })
  const res = await client.commit('default', {
    main: 'module.exports = () => {}'
  })

  t.deepEqual(res, {ok: 1})
})

test('Test refresh token', async t => {
  t.plan(3)

  nock('https://screeps.com')
    .matchHeader('content-type', 'application/json')
    .matchHeader('accept', 'application/json')
    .post('/api/auth/signin', {
      email: 'foobar',
      password: 'barbaz'
    })
    .reply(() => {
      t.pass()

      return [200, {
        ok: 1,
        token: 'quuxnorf'
      }]
    })

  const client = await new ScreepsModules({
    email: 'foobar',
    password: 'barbaz'
  })

  const token = await client.refreshToken('/test')
  t.is(token, 'quuxnorf')
  t.is(client.options.token, token)
})

test('Test token in request', async t => {
  t.plan(1)

  nock('https://screeps.com')
    .matchHeader('x-token', 'foobar')
    .matchHeader('x-username', 'foobar')
    .get('/test')
    .reply(() => {
      t.pass()

      return ok
    })

  const client = await new ScreepsModules({
    token: 'foobar'
  })

  await client.request('/test')
})

test('Test always consumes token from response', async t => {
  const client = new ScreepsModules()

  nock('https://screeps.com')
    .get('/token-response')
    .reply(200, 'Foobar', {'X-Token': 'foobar'})

  await client.request('/token-response')
  t.is(client.options.token, 'foobar')
})

test('Test removes token on 401 responses', async t => {
  const client = new ScreepsModules({
    token: 'foobar'
  })

  nock('https://screeps.com')
    .get('/unauthorized')
    .reply(401, 'Unauthorized')

  try {
    await client.request('/unauthorized')

    t.fail()
  } catch (e) {
    t.is(client.options.token, '')
  }
})

test('Test custom server URL', async t => {
  t.plan(1)

  nock('http://localhost:8888')
    .post('/foo/api/user/code')
    .reply(() => {
      t.pass()

      return ok
    })

  await new ScreepsModules({
    serverUrl: 'http://localhost:8888/foo'
  }).commit()
})

test('Test custom server password', async t => {
  t.plan(1)

  nock('http://localhost:8888')
    .matchHeader('X-Server-Password', 'foobar')
    .post('/api/user/code')
    .reply(() => {
      t.pass()

      return ok
    })

  await new ScreepsModules({
    serverUrl: 'http://localhost:8888/',
    serverPassword: 'foobar'
  }).commit()
})

test('Test commit without branch', async t => {
  t.plan(1)

  nock('https://screeps.com')
    .post('/api/user/code', {
      modules: {foo: 'bar'}
    })
    .reply(() => {
      t.pass()

      return ok
    })

  await new ScreepsModules().commit({
    foo: 'bar'
  })
})

test('Test commit gzip', async t => {
  t.plan(2)

  nock('https://screeps.com')
    .matchHeader('content-encoding', 'gzip')
    .matchHeader('content-type', 'application/json')
    .post('/api/user/code')
    .reply((req, body) => {
      t.deepEqual(body, {
        modules: {
          foo: 'bar'
        }
      })
      t.pass()

      return ok
    })

  await new ScreepsModules({gzip: true})
    .commit({foo: 'bar'})
})

test('Test retrieve modules', async t => {
  t.plan(2)

  nock('https://screeps.com')
    .get('/api/user/code')
    .reply(() => {
      t.pass()

      return {
        main: 'module.exports = () => {}'
      }
    })

  const client = await new ScreepsModules()
  const res = await client.retrieve()

  t.deepEqual(res, {main: 'module.exports = () => {}'})
})

test('Test retrieve modules from branch', async t => {
  t.plan(1)

  nock('https://screeps.com')
    .matchHeader('accept-encoding', 'gzip, deflate')
    .get('/api/user/code')
    .query({branch: 'foobar'})
    .reply(() => {
      t.pass()

      return ok
    })

  const client = new ScreepsModules()
  await client.retrieve('foobar')
})

test('Test bad response', async t => {
  t.plan(2)

  nock('https://screeps.com')
    .get('/foo')
    .reply(() => {
      t.pass()

      return [400, 'Bad request']
    })

  const client = new ScreepsModules()

  try {
    await client.request('/foo')

    t.fail()
  } catch (e) {
    t.truthy(e.toString().match(/Bad request/))
  }
})

test('Test error in response', async t => {
  const client = new ScreepsModules()

  nock('https://screeps.com')
    .matchHeader('accept', 'application/json')
    .get('/bar')
    .reply(() => {
      t.pass()

      return [400, {error: 'Bad response'}]
    })

  try {
    await client.request('/bar', {method: 'GET', json: true})

    t.fail()
  } catch (e) {
    t.is(e.toString(), 'Bad response')
  }

  nock('https://screeps.com')
    .matchHeader('accept', 'application/json')
    .get('/baz')
    .reply(() => {
      t.pass()

      return [200, {error: 'Bad response'}]
    })

  try {
    await client.request('/baz', {method: 'GET', json: true})

    t.fail()
  } catch (e) {
    t.is(e.toString(), 'Bad response')
  }
})

test('Test aliases', t => {
  t.plan(5)

  class TestClient extends ScreepsModules {
    commit (...args) {
      t.deepEqual(args, ['foo', 'bar'])
    }

    retrieve (...args) {
      t.deepEqual(args, ['bar', 'baz'])
    }
  }

  const client = Object.create(TestClient.prototype)

  client.fetch('bar', 'baz')

  client.up('foo', 'bar')
  client.down('bar', 'baz')

  client.push('foo', 'bar')
  client.pull('bar', 'baz')
})


index.js
const debug = require('debug')('screeps-modules')
const request = require('request')
const zlib = require('zlib')

module.exports = class ScreepsModules {
  constructor (options) {
    this.options = Object.assign({}, {
      email: '',
      password: '',
      token: '',
      serverUrl: 'https://screeps.com',
      serverPassword: '',
      gzip: false
    }, options)
  }

  refreshToken () {
    const {email, password} = this.options

    return this.request('/api/auth/signin', {
      method: 'POST',
      json: {
        email,
        password
      }
    })
    .then(({ok, token}) => {
      if (ok) {
        this.options.token = token

        return token
      }
    })
  }

  commit (branch, modules) {
    if (!modules) {
      modules = branch
      branch = undefined
    }

    if (!this.options.gzip) {
      return this.request('/api/user/code', {
        method: 'POST',
        json: {
          branch,
          modules
        }
      })
    }

    const compressed = zlib.gzipSync(JSON.stringify({
      branch,
      modules
    }))

    return this.request('/api/user/code', {
      method: 'POST',
      headers: {
        'Content-Encoding': 'gzip',
        'Content-Type': 'application/json'
      },
      body: compressed.toString('hex')
    })
  }

  retrieve (branch) {
    return this.request('/api/user/code', {
      method: 'GET',
      gzip: true,
      json: true,
      qs: {
        branch
      }
    })
  }

  fetch (...args) {
    return this.retrieve(...args)
  }

  up (...args) {
    return this.commit(...args)
  }

  down (...args) {
    return this.retrieve(...args)
  }

  push (...args) {
    return this.commit(...args)
  }

  pull (...args) {
    return this.retrieve(...args)
  }

  request (url, options = {}) {
    options.uri = url
    options.auth = this.auth(url, options)
    options.baseUrl = this.options.serverUrl

    this.auth(url, options)

    debug(`Requesting: ${JSON.stringify(options, null, 2)}`)

    return new Promise((resolve, reject) => {
      request(options, (err, response, body) => {
        debug(`Error: ${JSON.stringify(err)}`)
        debug(`Response: ${JSON.stringify(response)}`)
        debug(`Body: ${JSON.stringify(body)}`)

        if (err) {
          return reject(err)
        }

        if (response.statusCode === 401) {
          this.options.token = ''
        }

        if (response.headers['x-token']) {
          this.options.token = response.headers['x-token']
        }

        if (body.hasOwnProperty('error')) {
          return reject(body.error)
        }

        if (response.statusCode % 400 < 200) {
          return reject(body)
        }

        resolve(body)
      })
    })
  }

  auth (url, options) {
    if (url === '/api/auth/signin') {
      return
    }

    const {email, password, token, serverPassword} = this.options

    if (token !== '') {
      options.headers = Object.assign({}, options.headers, {
        'X-Token': token,
        'X-Username': token
      })
    } else {
      options.auth = {
        username: email,
        password
      }
    }

    if (serverPassword !== '') {
      options.headers = Object.assign({}, options.headers, {
        'X-Server-Password': serverPassword
      })
    }
  }
}
